predicthq/kafka/consumer.py

```diff
@@ -0,0 +1,163 @@
+        for partition, m_partition in itertools.groupby(m_topic, lambda x: x.partition()):
+            tp = Tp(topic, partition)
+            if tp not in batch_ref:
+                batch_ref[tp] = []
+            batch_ref[tp] += list(m_partition)
+    return batch_ref
+
+
+class PhqKafkaConsumer(object):
+    def __init__(
+        self, input_topic: str, group_id: str, batch_size: int, consumer_timeout_ms: int,
+        kafka_bootstrap_servers: List[str], kafka_consumer_config: Dict[str, str] = None
+    ):
+        self.closed = False
+
+        self.input_topic = input_topic
+
+        self._consumer_timeout_ms = consumer_timeout_ms
+
+        self._consumer_group = group_id
+        self._consumer_batch_size = batch_size
+        self._consumer = get_kafka_consumer(
+            kafka_bootstrap_servers,
+            self._consumer_group,
+            kafka_consumer_config
+        )
+
+        self._consumer.subscribe([self.input_topic])
+
+    def _format_message_batch(self, kafka_messages):
+        messages = []
+        batch_ref = {}
+        for tp, batch in _group_messages(kafka_messages).items():
+            topic = tp.topic
+            partition = tp.partition
+
+            if topic == self.input_topic:
+                if topic not in batch_ref:
+                    batch_ref[topic] = {}
+
+                for message in batch:
+                    # Skip null messages - they are used to 'delete' keys from
+                    # compacted topics.
+                    if message.value() is None:
+                        continue
+
+                    if message.error():
+                        error = message.error()
+                        if error.code() == confluent_kafka.KafkaError._PARTITION_EOF:
+                            continue
+                        else:
+                            raise confluent_kafka.KafkaException(msg.error())
+
+                    payload, ref = unpack_kafka_payload(message)
+                    message_id = ref['key']
+                    messages.append(Message(message_id, payload, ref))
+
+                batch_ref[topic][partition] = (batch[0].offset(), batch[-1].offset())
+
+            else:
+                log.warning('Received %(batch)s messages for unknown topic %(unknown_topic)s (expected %(expected_topic)s). Skipping.',
+                            {'batch': len(batch), 'unknown_topic': topic, 'expected_topics': self.input_topics})
+                continue
+
+        return messages, batch_ref
+
+    def _get_kafka_messages(self):
+        while True:
+            # The consumer interface doesn't obey the 'consumer_timeout_ms' consumer config.
+            log.debug('Calling consumer.consume, timeout %(time_s)s',
+                      {'time_s': self._consumer_timeout_ms})
+            messages = self._consumer.consume(num_messages=self._consumer_batch_size, timeout=self._consumer_timeout_ms / 1000)
```

Comment:
how do they honor the threshold? either or?

Reply:
I'm not sure I understand the question? basically, It will consume num_message if possible, if at the end of the timeout num_message isn't reach, it will return the current batch (0 to num_message).